### ISSUE TITLE
Add optional support for generic injection

### ISSUE THREAD

**Issue by dalewking** [google/dagger#213]
One of the problems with using Dagger 2 (particularly on Android) is the need to call a different method for every class that needs injection. E.g. see this article which speaks of it as an annoyance: http://blog.gouline.net/2015/05/04/dagger-2-even-sharper-less-square/.

People have handled this in different ways. Some just ignore it and have the class that needs injection bound at compile time to a particular component which to me means you are not actually doing DI because you cannot change your dependencies at run time.

But I think I have a way that Dagger could add a small feature that would make this all so much easier. The feature would be optional and if you don't want to use it would have no impact on you.

One of the ways that has been used to solve it is to use reflection and scan all the methods on the component to find inject methods and create a map from class to method. But it occurs to me that instead of resorting to reflection to create this map, it would be much handier if Dagger 2 would build this mapping for the user since it knows at compile time what the mapping is.

 So what I propose is this: In addition to the other Component methods you allow another method signature:

    public abstract <T> MembersInjector<T> someMethodName(Class<T> c);

The presence of that method signature will be what controls this optional feature. When present the component generated will have the following added to it:

 - A field containing a Map&lt;Class&lt;\*&gt;, MembersInjector&lt;\*&gt;&gt;
 - The initialize method will generate code to fill the map which would basically just be a bunch of puts
 - The above method signature implementation would return a lookup from the map.

That is the minimum level of support needed but it would be even nicer to allow a generic inject method as in:

    public abstract <T> void someMethodName(T instance);

This method would also trigger creating the mapping. The implementation would look up in the map for the class of the instance to get an injector for doing injection. If it doesn't find one then it tries the superclass and so on up until it gets to java.lang.Object.

What does it do if there is no injector for the given type? It should not throw an error. In the case of the return type being void it should do nothing. If the return type is T it should return null. You could also allow a boolean return value to indicate if injection succeeded. I suppose you could also declare an InjectionNotFound exception and if the method is declared to throw it then you throw the exception, but if it isn't declared, don't throw the exception.

**Comment by rgrinberg** [comment]
+1 for some way to address this problem. This has been my biggest issue when migrating from dagger 1 to dagger 2.

**Comment by tbroyer** [comment]
https://github.com/tbroyer/bullet/ ?

**Comment by dalewking** [comment]
Was not aware of Bullet. API-wise Bullet is the kind of thing I am after. Your internal implementation of a linear search with a bunch of if-instanceof statements is not acceptable however. The correct implementation is to build a map from class to implementation so injection is fast.

I was contemplating writing a plugin like yours when I decided it would be a lot better if this capability was built into Dagger since Dagger already has all the information instead of another plugin trying to figure out the information from what dagger generated.

So I think my request is still valid that this can and should be built into Dagger.

Meanwhile I might try forking Bullet to provide a better implementation using a map.

**Comment by tbroyer** [comment]
@dalewking Would a "switch on strings" be better? (on `getCanonicalName()`, looping over `getSuperclass()` starting from the instance's `getClass()`)
IIRC, Dagger 1 was built around strings to avoid loading too many classes (eagerly) as it could be a performance issue on Android, so a switch-on-strings could likely help here.

The reason it's a linear search is that I don't expect that many cases in the if-cascade, and there shouldn't be that many calls to it either; so I don't expect it to be an issue in practice. But improvements (or suggestions of improvements) are welcome. Also, indeed feel free to fork and make your own (I'd be grateful if you could give me some feedback then).

**Comment by dalewking** [comment]
Switch on Strings is a very bad idea as it prevents obfuscation. I can tell you that in our application we have a component with 54 inject statements for all the injectable activities, fragments, and views, so worst case of 54 instanceOf calls which is not acceptable.

I put my response onto the Bullet issues list: https://github.com/tbroyer/bullet/issues/2 rather than spamming this list.

**Comment by netdpb** [comment]
What in your opinion is most annoying about having a different `inject(Foo)` method for each concrete type `Foo`?

* Writing all those methods?
* Managing the dependencies from the component to all those concrete types?
* Something else?

**Comment by dalewking** [comment]
@netdpb If you were just doing straight Java (not Android) it wouldn't be such a big deal, but on Android it is a big deal and actually makes it very difficult to actually do dependency injection.

In Android the majority of the things you need injected you cannot create (e.g. Activity) therefore you have to actually request injection from the object itself. There are many Dagger 2 examples out there where they are using Dagger 2, but aren't in my opinion actually doing dependency injection!

Dependency injection means that at runtime you can inject different things into the class being injected. If the class being injected is compile-time bound to one particular component then you are not doing DI. You are no better than if you just instantiated your dependencies inside the class.

There are examples where people actually create the component to do the injection INSIDE the object being injected. That binds at compile time what will be injected and is no better than just instantiating your dependencies inside the class. Some people hide that in a base class but that really isn't any better as you are still hardcoding at compile time what is going to be injected.

There are ways to alleviate the problem but they usually involve lots of boilerplate in the classes being injected (see annoyances #1 in this article for an example: https://blog.gouline.net/2015/05/04/dagger-2-even-sharper-less-square).

The fact that you have to bind at compile time to the injection method makes this harder to solve.

**Comment by JakeWharton** [comment]
How does your proposal solve the problem of the fact that it breaks full graph analysis? Dagger 1 had the `injects=` list and Dagger 2 has methods on the component to bridge the gap. Your proposal lacks anything which would allow tying injection sites to modules at compile time which turns it into a just-in-time binding at runtime that can fail.

**Comment by netdpb** [comment]
I don't think that requiring the component know about the types it needs to inject at compile time disqualifies Dagger from being DI.

Even if `MyActivity` contains a nested component, it can be injected by any component with an `inject(MyActivity)` method. There's nothing special about the nesting relationship in this case; it's just convenient.

And `MyActivity` still doesn't have to know anything about the implementations or creation strategies for its dependencies, which is the real point of DI.

The point of Dagger is to resolve the entire dependency graph at compile time. So you'll never get rid of the compile-time dependency from the type that is actually annotated with `@Component` and all of the implementations that it uses to create the object graph.

**Comment by dalewking** [comment]
The problem isn't the component knowing the types it needs to inject. That is fine and necessary. The problem is the object being injected knowing, or more precisely being bound at compile time to, one specific component. If the object being injected is bound at compile time to the component, then you can't substitute a different component.

Dependency injection is about the ability to substitute different implementations at RUNTIME for the items to be injected (e.g. when running in production and when running in a unit test). Tying injection sites to modules at COMPILE TIME is incompatible with that, therefore it isn't doing dependency injection since what is injected is determined at compile time.

So the fact that MyComponent has the method inject(Foo) is necessary. But if Foo itself is bound at compile time in any way or has any knowledge whatsoever (direct or indirect) about the existence of MyComponent that is a problem. It prevents you from easily being able to instead use MyUnitTestComponent for injection.

The way I interpret the notion of "full graph analysis" is the root of the problem here. Dagger guarantees that if a component has an inject(Foo) method then calling that method will succeed at runtime (barring exceptions being thrown in the module) and that can be checked at compile time. No problem there. And the proposal does not change that.

But there is this other myth that has sprung up around Dagger 2 (which I think is what you are calling full graph analysis) that you should be able to determine at compile time that there is a component that can inject Foo and that is THE component that will do the injection. This is incompatible with actually doing DI, since we shouldn't even know until run time what component will do the injection. If we know at compile time what component will do the injection then you cannot use a different component at run time.

So yes with this proposal if you chose a component at run time that did not have the inject(Foo) method it will fail at run time. There is no way to prevent that and it cannot be determined at compile time since you cannot know at compile time what component will be used to inject Foo.

So @netdpb to be clear the proposal does not eliminate the inject(Foo) methods or MembersInjector<Foo> getFooInjector methods on the component. It just allows you to request an additional generic method that will accept and invoke the correct method (it would actually use the intermal MembersInjector variables rather than invoking the methods).

So @JakeWharton the inject methods on the component are still there which are the equivalent of injects lists on modules in dagger 1. The proposal is to allow the generic inject method of Dagger 1. There are libraries like Bullet (https://github.com/tbroyer/bullet) and my rewrite of Bullet (https://gitlab.com/NobleworksSoftware/NobleInjection/tree/master) which create classes that can wrap a component with class with the generic method, but doing it in the another class is much less efficient than if it were done in the component itself.

And this is proposed as optional. If you want to keep pretending you are actually doing DI when using Dagger 2 bound at compile time to the component, then the proposal does not affect you.

And note

**Comment by JakeWharton** [comment]
> The problem is the object being injected knowing, or more precisely being bound at compile time to, one specific component.

This is a user's choice. I fail to see how specifying a field injection on a component tells the object anything about how it's being injected.

**Comment by netdpb** [comment]
I don't understand this part:

> The problem is the object being injected knowing, or more precisely being bound at compile time to, one specific component.

The practice of nesting a component that knows how to inject `Foo` into `Foo` itself is a convenience, and as @JakeWharton says, the user's choice. It doesn't prevent some other component from also having an `inject(Foo)` method, and it doesn't give that nested component any special knowledge about `Foo` (or _vice versa_).

> There is this other myth that has sprung up around Dagger 2…that you should be able to determine at compile time that there is a component that can inject `Foo` and that is THE component that will do the injection. This is incompatible with actually doing DI.

I suppose hard-linking an implementation to the component that must provide it would be bad, but I don't agree that Dagger 2 encourages that. I think Dagger 2 requires that _users of the component type_ must know at compile time what classes they'll want the component to provide and inject, but that's different from _implementations provided by the component._ That is one of the principal (and principled) differences between Dagger 2 and Dagger 1.

**Comment by tbroyer** [comment]
@netdpb Yet this is what the sample does, creating a `DaggerHomeComponent` that couldn't be switched for anotger during tests: https://github.com/google/dagger/blob/master/examples/android-activity-graphs/src/main/java/com/example/dagger/activitygraphs/ui/HomeActivity.java

**Comment by JakeWharton** [comment]
That sample was written by me the night before Dagger 1.0 was released in a frenzy to appease (and anticipate) Android people wanting copy/pasta examples. In the course of 9 days, me, Jesse, and others [released 7 major versions][1] of open source projects. To say that sample was rushed and lacked any and all forethought is an understatement.

 [1]: https://corner.squareup.com/2013/05/seven-days-of-open-source.html

**Comment by JakeWharton** [comment]
Regardless, I firmly believe activities should inject themselves anyway, and you should use "proper" DI on everything else in your graph. u2020 does plenty of runtime dependency switching while still having an activity that injects itself. Proper use of composition in activities rather than loading them up as an actual UI god object goes a long way to fix this problem.

**Comment by dalewking** [comment]
> The practice of nesting a component that knows how to inject Foo into Foo itself is a convenience, and as @JakeWharton says, the user's choice. It doesn't prevent some other component from also having an inject(Foo) method, and it doesn't give that nested component any special knowledge about Foo (or vice versa).

The trouble is that all of the Android examples out there make that choice and it is very difficult to make any other choice.

>I suppose hard-linking an implementation to the component that must provide it would be bad, but I don't agree that Dagger 2 encourages that. I think Dagger 2 requires that users of the component type must know at compile time what classes they'll want the component to provide and inject, but that's different from implementations provided by the component. That is one of the principal (and principled) differences between Dagger 2 and Dagger 1.

It actually does encourage that as it is the easiest way to implement it. So the component has an inject(Foo) method, the question is how does that method get called? Java does not have double dispatch so the call to the inject method must be made from a call that has a COMPILE time type of Foo. If the component itself is the one that declares the inject method how can you make the call without a reference of the type of the actual component? At the bare minimum you should have an interface that declares the inject method and the component just implements that interface. Then you can at least make a call with only a reference to the interface. This at least is a way to isolate Foo from the actual type of the component. This however requires some boilerplate in each class being injected. Usually this boilerplate would be a method that each class that needs injection implements that accepts an instance of the interface and each class has an implementation that looks exactly the same like:

@Override
protected void injectComponent(InjectionGraph graph)
 {  
    component.inject(this);
}

But this would all be much simpler if a component can simply implement an interface like this:

public interface Injector
{
    <T> void inject(T instance);
}

**Comment by JakeWharton** [comment]
There's no reason you can't make your components implement an interface for swapping between them. It would have to be strongly-typed on the injectee, obviously, since Dagger needs that information at compile-time. Your single inject method can't work without a reflective implementation, which seems exactly what Bullet provides.

**Comment by dalewking** [comment]
> It would have to be strongly-typed on the injectee, obviously, since Dagger needs that information at compile-time.

Why? That is incorrect. Dagger does not NEED that information at all. I agree with the current implementation it has to be strongly typed because of this problem with Dagger 2 and that is what needs to change.

> Your single inject method can't work without a reflective implementation, which seems exactly what Bullet provides.

You have to be precise with what you mean by reflection. The proposal, Bullet, and my rewrite of Bullet do not reference anything in java.lang.reflect, do not have slow performance, and do not prevent obfuscation. The only thing close to reflection they rely on are getClass() and in some situations getSuperclass() which are technically not reflection. It does determine the correct course of action based on the run time type of the instance but that is not the same as reflection.

**Comment by JakeWharton** [comment]
> Dagger does not NEED that information at all.

It does for full graph analysis which pretty much the entire premise Dagger was founded on (aside from avoiding reflection).

How can Dagger match the dependencies requested by injection sites and dependencies provided by modules in your API without an explicit bridge between a list of modules+components and types being instance-injected (not constructor injected)?

> technically not reflection

I keep telling people Dagger 1 technically didn't do reflection, only `Class.forName`. That's just class loading, after all, which happens all the time!

**Comment by dalewking** [comment]
> It does for full graph analysis which pretty much the entire premise Dagger was founded on (aside from avoiding reflection).

To respond to that I would like to see a definition of the "full graph analysis" you keep referring to. I think I know what it means and that the notion that it is compatible with actual DI is wrong, but I don't want to argue against a straw man.

The graph analysis that Dagger does is that if a component has a method like inject(Foo) then there are providers for all dependencies. That is the only graph analysis done by Dagger and this proposal changes none of that. To be clear I am not suggesting components would remove any of the inject methods that they would have now, just that on top of them you can add a generic injection method that looks up the right implementation (more precisely for the component it creates a map from class to MembersInjector and the generic method simply gets the instance corresponding to the class (see this file from my library for an idea of how simple the generated code would be https://gitlab.com/NobleworksSoftware/NobleInjection/blob/master/generic-component-runtime/src/main/java/com/nobleworks_software/injection/GenericComponent.java)

> How can Dagger match the dependencies requested by injection sites and dependencies provided by modules in your API without an explicit bridge between a list of modules+components and types being instance-injected (not constructor injected)?

It will still have the bridge from component to the classes being injected in that for any type that is instance injected you will have to define an inject method on the component. That does not change.

However there would not be, nor should there be, a link from the type being instance injected to the actual component that does the injection. If you have such a link, particularly one defined at compile time, this is where I claim you would not actually be doing DI.

> I keep telling people Dagger 1 technically didn't do reflection, only Class.forName. That's just class loading, after all, which happens all the time!

That actually could properly be called reflection but of course the real problem was that it was not compatible with obfuscation which is more important than quibbling over what the definition of reflection is.

Perhaps it would be clearer if you looked at my library which uses annotation processing to generate wrapper classes for Dagger components. For every injection method in the component it put an entry in the map from class to injector.

The place to start is the base class for the wrappers it generates: https://gitlab.com/NobleworksSoftware/NobleInjection/blob/master/generic-component-runtime/src/main/java/com/nobleworks_software/injection/GenericComponent.java.

That class pretty much tells the whole story. The classes generated extend this and basically just override the buildInjectorMap method to build the map. Unfortunately, how to create that map is a bit complicated because I don't have access to the internals of the component. If this were done by Dagger itself it would be very straightforward (just a bunch of calls to put).

There is the ClassInjectorMap which is just a highly optimized hash map to take advantage of the fact that the map will hold a known number of items and we never delete anything from it.

The proposal would be that any component that defines a generic injection method would extend a class similar to this and the component would populate the map.

**Comment by JakeWharton** [comment]
> just that on top of them you can add a generic injection method that looks up the right implementation

What prevents me from calling this API with a type that is not listed concretely and therefore participated in the full graph analysis? That's a runtime failure which nobody wants.

**Comment by JakeWharton** [comment]
> compatible with obfuscation

Obfuscation is stupid. But the comment was a joke, anyway.

**Comment by dalewking** [comment]
> What prevents me from calling this API with a type that is not listed concretely and therefore participated in the full graph analysis? That's a runtime failure which nobody wants.

Nothing prevents that. The notion that you can prevent that at compile time is the idea I am trying to combat. If there is a way to detect that at compile time then you by definition have hard coded at compile time what component will do the injection and cannot choose the component at run time and thus are not actually doing DI.

Since with real DI the component is chosen at run time then the only place it can fail is at run time, so yes that is a run time failure I do want. If you want to live under the illusion that it is preventable at compile time then just don't use this extra feature.

Granted you can get some of the compile time checking without tying to the actual component using the superinterface that declares the inject methods at the cost of duplicate boilerplate code in every class that needs instance injection.

One problem I have with that is that if I create one interface for all things that need injected then I am tying every injectable type to that interface which in turn couples it to every other injectable type. In other words that interface refers to every injectable type and every injectable type refers to it and thus all are coupled together.

You can instead declare an injection interface per injectable type but then again you are adding even more boilerplate code to every injectable type.

Ask yourself seriously why you continue to use Dagger 1 over Dagger 2 as you said in the u2020 issue:

> It serves our use case much better than Dagger 2. We leverage granular overrides in the run time and its flexibility for testing that cannot be achieved in v2 without far too much ceremony.

That inflexibility in the run time is what I am talking about. Restoring the ability to inject generically  would get that flexibility back and remove the reasons you are staying with Dagger 1. I realize that the lack of Dagger 1 overrides is also a pain, but there are ways to work around that limitation for most cases.

**Comment by JakeWharton** [comment]
> The notion that you can prevent that at compile time is the idea I am trying to combat

Ok. I'm a giant :-1: for this feature being in Dagger 2.

> remove the reasons you are staying with Dagger 1

This would add a reason. It's a bad idea.

**Comment by netdpb** [comment]
Are you comfortable with having a compile-time list of the classes you want to be able to `inject()` somewhere, or must it be discovered by analyzing the component?

If you are, then what if you had an annotation that listed the types A, B, C you want to be able to inject, and it generated a class like this:

```java
class MultiInjectorImpl implements MultiInjector {
  private final MembersInjector<A> aInjector;
  private final MembersInjector<B> bInjector;
  private final MembersInjector<C> cInjector;

  @Inject MultiInjectorImpl(MembersInjector<A> aInjector, MembersInjector<B> bInjector, MembersInjector<C> cInjector) {
    this.aInjector = aInjector;
    this.bInjector = bInjector;
    this.cInjector = cInjector;
  }

  @Override
  public void inject(Object foo) {
    // In an order such that subtypes are tested before supertypes.
    if (foo instanceof A) aInjector.injectMembers((A) foo);
    else if (foo instanceof B) bInjector.injectMembers((B) foo);
    else if (foo instanceof C) cInjector.injectMembers((C) foo);
  }
}
```

and a module that bound `MultiInjector` to that implementation.

Then your component could install that module, and provide a `MultiInjector`. Would that get you what you're looking for?

**Comment by netdpb** [comment]
By the way, I agree that the amount of work you have to do to avoid having a huge component type that has compile-time dependencies on every concrete activity class in Android is too large. I do feel the pain that @dalewking is trying to alleviate.

I want there to be a good solution. I'm just not convinced that this solution is worth the cost of losing the compile-time type checking feature of Dagger 2.

**Comment by ENikS** [comment]
As I mentioned in #186 implementing **Inject(Object obj)** on each component implementation would solve issue of strong type requirement.

When graph is built during compilation and component implementation is generated, it already has all the required information. Reusing it to generate one more method is trivial. 

In order to limit number of instanceOf() calls processor could walk inheritance chain and generate case for each base class of the type. For interface like this:
```java
@Component...
interface Foo {
   Inject(type1 obj);
   Inject(type1subsub obj);
}
```
it will generate following switch:
```java
...
      switch(obj.getClass().getName()) { 

          case type1.class.getName():
          case type1sub.class.getName():
              inject((type1)obj);
              break;

          case type1subsub.class.getName():
              inject((type1subsub)obj);

          // All component injectors
          ...
      }
...
```

**Comment by dalewking** [comment]
> > The notion that you can prevent that at compile time is the idea I am trying to combat

> Ok. I'm a giant :-1: for this feature being in Dagger 2.

> >  remove the reasons you are staying with Dagger 1

> This would add a reason. It's a bad idea.

Your lack of logic defies explanation. You admit that you do not use Dagger 2 and instead use Dagger 1. Dagger 1 has a generic inject method and there is nothing preventing you from telling your Dagger 1 ObjectGraph to inject something that it does not know how to inject causing an exception at runtime. This is making Dagger 2 more like Dagger 1 in that respect. If it is such a bad idea why do you continue to use Dagger 1. I would argue that the main reason is that Dagger 1 works this way.

**Comment by JakeWharton** [comment]
> Your lack of logic defies explanation.

Your inability to see reasoning on the value of full compile-time graph validation and usage does not equate to a lack of logic. Making Dagger 2 more strict in this regard was a design requirement and motivation for its creation.

> This is making Dagger 2 more like Dagger 1 in that respect.

Which is a bad idea.

> If it is such a bad idea why do you continue to use Dagger 1.

Because I value things completed unrelated to this "feature" that it provides but Dagger 2 does not.

> I would argue that the main reason is that Dagger 1 works this way.

It isn't.

**Comment by ENikS** [comment]
Boys, measuring your members wouldn't take anyone far. 
How about we discuss solutions to the problem instead of personal logic?

**Comment by JakeWharton** [comment]
A solution was posted on July 14th.

**Comment by ENikS** [comment]
Could you point me to it?

**Comment by JakeWharton** [comment]
https://github.com/google/dagger/issues/213#issuecomment-121155656

**Comment by netdpb** [comment]
I agree: Please treat each other as if we are all trying to solve problems, even if we disagree about what the problems actually are. (Because it's true.)

**Comment by dalewking** [comment]
@netdpb What you are describing is similar to what Bullet and my refactoring of Bullet are basically doing. They create wrappers of a component so instead of passing member injectors you accept a Component. Bullet does what you show with a big if-else chain checking instance of but that is too inefficient. We have a component with 54 injection methods so that would be a bad idea. The point of my rewrite of Bullet was to replace the if-else chain with a lookup in a map.

So yes, my library (https://gitlab.com/NobleworksSoftware/NobleInjection) generates classes like this. The problem is that without access to the internals of the component it has to do some very hacky things that would not be necessary if this were built into the Dagger component and would be much more efficient.

And it is a misnomer to say you are losing the compile time type checking. It is an optional feature so you are allowing it to be disabled. But as I keep stressing the notion that it is really possible when doing injection into an instance is an illusion.

**Comment by ENikS** [comment]
@dalewking  So, I guess adding **Inject(Object obj)** to the component and **switch** based approach to resolve type instead of **instanceOf()** should provide acceptable solution for this problem? 

As I see it:
1. will be able to use injection within base class implementation without requirement to provide strongly typed reference. (or anywhere else for that matter)
2. will still be generating everything at compile time.
3. will be compatible with existing API so strongly typed injectors could be used
4. code could still be obfuscated if required (valid requirement btw)

Adding extra method to the component using already collected graph/type info should be relatively trivial.

@netdpb How would you suggest we approach this? Should we submit a feature request and Google will handle it in a reasonable time frame or should we try to implement it ourselves and just submit a pull request?

**Comment by dalewking** [comment]
So the thing you are claiming is a solution is Bullet. As I said, earlier API-wise bullet is exactly what we are after, creation of a generic inject method, but the implementation of Bullet is unacceptable due to the O(n) search on every injection due to the if-else chain.

I have my library that is a rewrite of the injection part of Bullet which fixes the implementation by using a hash map which gives O(1) performance for injection by adding a one time O(n) setup of the map. But as I said any 3rd party library cannot do this as efficiently as can be done inside of Dagger. So the proposal is to basically do what these libraries are doing but make it part of Dagger so that it can be implemented in an efficient and straightforward manner.

Another alternative would be to open up the generated component to allow access to the internals so that a third party solution can do this. For example, if the generated component was not final allowing it to be subclassed and all the member injector fields were declared protected then I could rewrite my library to do this as efficiently as could be done inside Dagger. Or perhaps just declaring them with package level access.

Or do what I called the bare minimum implementation which is to support generation of an implementation of a generic members injector method like this:

     <T> MembersInjector<T> someMethodName(Class<T> c);

That implementation is simply looking up in a map (see my library for a more optimized implementation than using a standard HashMap). The other part is constructing the map which is essentially just a lazy construction of the map which is just a bunch of put statements in the component.

I really don't see what is so objectionable about this. If you don't like it don't use it, its optional. In my opinion something like this is vital for using Dagger 2 with Android.

I've thought about trying to do a PR for Dagger 2 to support this, but I really could not make much headway in understanding how Dagger 2 is implemented.

**Comment by JakeWharton** [comment]
I don't see how this is any different than assisted injection which [already lives][1] in a separate repo that composes with Dagger 2 just fine.

No one has made a compelling argument as to why this needs to be a feature that lives in Dagger 2 itself.

 [1]: http://github.com/google/auto/

**Comment by ENikS** [comment]
Well, how about to be able to utilize polymorphism in my code? It is one of the fundamental features of the Java language and at present Dagger 2 lucking any support for it. This simple method will add necessary support.

**Comment by dalewking** [comment]
@ENikS I agree implementation is trivial. But a switch statement is not an acceptable implementation. First off what are you going to use as the argument to the switch statement? TBroyer already proposed doing a switch on class name which I reject for 3 reasons:
 - It prevents obfuscation using Proguard 
 - It is even less efficient than the if-else-instanceOf chain
 - It requires Java 7, which is a problem for Android which still requires Java 6 support to support older devices.

The only acceptable solution that I have found is the creation of a hash map from Class to MembersInjector (but see my optimized implementation of the map using open hashing instead of relying on HashMap https://gitlab.com/NobleworksSoftware/NobleInjection/blob/master/generic-component-runtime/src/main/java/com/nobleworks_software/injection/ClassInjectorMap.java)

**Comment by JakeWharton** [comment]
> This simple method will add necessary support...

It's trivial to add assisted injection in the same manner, though. And unlike this proposal, can actually be done in a type-safe way with compile-time verification of the graph.

**Comment by ENikS** [comment]
@dalewking  I am not familiar with bytecode generated for switch in Java. Why do you think it is less efficient than if else statements?  In C# and C++, as I think in any descent VM it should be very quick, one step jump. 

@JakeWharton If you say it is so trivial, perhaps google could provide example of how it should be handled properly? And what about polymorphism? 
I honestly don't see much value in a library which required 10 more projects just to do basic things.

**Comment by JakeWharton** [comment]
> perhaps google could provide example of how it should be handled properly

No one said Google would provide a solution to this use case. Why do they? As I've said multiple times now, complete graph analysis and verification at compile-time was a core design requirement of Dagger 2 which this unequivocally violates. As a non-Google employee, this question has no relevance being directed towards me.

If you want to get into a realm where Google needs to somehow "bless" use cases, then Guice 4.0 is a fantastic example of a mature, fast, and full-featured library that provides runtime dependency-injection for injectee classes which need to be just-in-time injected.

> And what about polymorphism? 

I don't care. I'm not trying to solve this use-case. I'm trying to eliminate it as something that's Dagger's responsibility. It's not. This runtime version of injection is someone else's problem, not Dagger's.

**Comment by dalewking** [comment]
@ENikS So were you talking about doing a switch based on the class name?

OK, so switch on string is more efficient than I thought so is probably not worse than the if-else chain (see http://javarevisited.blogspot.com/2014/05/how-string-in-switch-works-in-java-7.html) but this solution is still unacceptable for the other reasons I mentioned.

**Comment by dalewking** [comment]
I have read about assisted injection by which I think you are referring to Googles AutoFactory (your link just pointed to the google/auto project). But I fail to see how this is in any way related to the subject at hand or how this is an alternative solution to the problem.

**Comment by JakeWharton** [comment]
>  I think you are referring to Googles AutoFactory (your link just pointed to the google/auto project)

Yes! Factory generation (via [auto/factory][1]) is done irrespective of the injector of choice. It's a fantastic mixin project which needs no knowledge of the larger injection libraries at play or decisions like those being made in this issue.

Dagger 1? Dagger 2? Guice? auto/factory cares not.

> But I fail to see how this is in any way related to the subject at hand or how this is an alternative solution to the problem.

Because it is a solution to a problem that a limited subset of Dagger users have in a way that can be done without modification of Dagger itself. Or the specificity of an injector.

Unequivocally, I do not need what you are proposing here. In fact, most likely do not as well. This is evidenced and corroborated by the fact that none such issue was ever filed before yours, and none run to your aid here. You are a specialized subset of users to which features mush be justified–rightfully so, but not wrongly so.

Since your proposal is fundamentally at odds with a core design principal of Dagger 2: the fact that graphs should be entirely linkable at compile-time (and thus statically resolvable and analyzable), how does a feature like what you propose fit in to the offering? It's _only_ something that exists as a mixin, and not something that's core the library itself.

 [1]: https://github.com/google/auto/tree/master/factory

**Comment by dalewking** [comment]
> Because it is a solution to a problem that a limited subset of Dagger users have in a way that can be done without modification of Dagger itself. Or the specificity of an injector.

But Autofactory itself does not solve the problem I am describing so is not relevant. Your only point seems to be that it was done as an add-on instead of changing Dagger. I already admit that it is "possible" to solve this without changing Dagger as TBroyer's Bullet library and my rewrite of it have done, but the problem is that they cannot solve it in an efficient manner.

> Unequivocally, I do not need what you are proposing here.

LOL! You aren't using Dagger 2! I tend not to need features implemented in libraries I don't use either! You ARE using Dagger 1 which has this feature so you are currently using this feature. When you actually use a DI framework that doesn't have this feature, then your statement might have some credibility.

 I have already suggested that the lack of this in Dagger 2 is a primary cause of why you and many others are sticking with Dagger 1. You said there were other features of Dagger 1 but did not elaborate. I cannot fathom what those other features are other than the lack of module overrides (which is cumbersome but can be worked around in most cases https://github.com/google/dagger/issues/186#issuecomment-163338603). Other than these 2 features I know of nothing that would make Dagger 1 a better solution than Dagger 2.

> In fact, most likely do not as well. This is evidenced and corroborated by the fact that none such issue was ever filed before yours, and none run to your aid here.

Quoting from early on in this issue:

> > +1 for some way to address this problem. This has been my biggest issue when migrating from dagger 1 to dagger 2.

And there is also the fact that TBroyer created Bullet which addresses this issue.

And there is also #110 and #186 (filed before mine) which basically talks about how it is very difficult to make Dagger 2 work with Android such that you can actually do unit testing. This issue is in my opinion the primary underlying cause of that.

And as I said in my first post, "see this article which speaks of it as an annoyance: http://blog.gouline.net/2015/05/04/dagger-2-even-sharper-less-square/". His solution is to add the boilerplate to every type that needs injection.

As circumstantial evidence I can point to the poor quality of all of the Dagger 2 examples out there and the fact that it appears that Dagger 2 is not really being adopted.

Granted, I am probably one of the few who have recognized that this issue is at the root of the problem. Many people like those in #110 and #186 have tried to use Dagger 2 and when they ran into problems thought the issue was a misunderstanding on their part instead of a problem with a core design principle of Dagger 2.
 
> Since your proposal is fundamentally at odds with a core design principal of Dagger 2: the fact that graphs should be entirely linkable at compile-time (and thus statically resolvable and analyzable), how does a feature like what you propose fit in to the offering?

In the case of a straight Java program where your sole interaction with a component is calling get in your main method this design principle works well.

But in cases like Android where someone else creates the objects you need to inject and an object that needs injecting itself has to do something to cause itself to be injected that design principle doesn't really work. It encourages tight coupling to the component itself (see the majority of Dagger 2 Android examples out there) and/or requires Herculean effort and a mound of boilerplate code to be added to the types to be injected.

So that design principle in the case of Android is at odds with ability to write loosely-coupled, cleanly-architected code. Do we hold onto that design principle and enforce that all users must abide by it no matter how much pain and bad code it causes or do we give people the option to relax that design principle to allow better coding with the trade-off that the injection may fail at run time?

> It's only something that exists as a mixin, and not something that's core the library itself.

And if it could be done as efficiently as a mixin I might agree although I would argue that in cases like Android the lack of this feature is encouraging very bad practices leading to highly coupled code and difficulty in writing unit tests.

**Comment by tbroyer** [comment]
One thing I don't understand @dalewking is whether you're concerned about the coupling with the component interface or the component (generated) implementation.
If you're concerned about the latter, then I fail to see how a “generic” non-typesafe method solves it.
If you're concerned about the former, then a) the component interface is _just_ an interface so it's IMO a non-issue (you can still choose the actual implementation at runtime), and b) there are ways around it that keeps it typesafe (splitting the interface into smaller interfaces). Yes those add more boilerplate, but it's all (and only) about how you want to reason about your code.

Your activity (the same would be true of a servlet BTW) “knows” it needs to be injected (it has `@Inject` annotations), and it “knows” its lifecycle is managed by something that won't do the injection, so it “knows” it has to inject itself. Being coupled to an interface that's capable of injecting itself is thus unavoidable, as well as the 3 lines of code to call that interface at the right time to inject itself.
Now's the time to decide the _scope_ of that interface, and every developer will have its own preferences (have one single-method interface per activity –which could be generated!–, group a few activities together, or have a single interface for the whole app).
If you want to trade compile-time typesafety (i.e. possible runtime error) for less boilerplate, that's your choice, but I, for one, totally agree with @JakeWharton that this should not be built into Dagger 2. Jake uses Dagger 1 –well, sure, he and his teammate @swankjesse built it– which works that way, so he knows from experience what he doesn't like about it (BTW, I may be wrong but I believe Jake worked with Google to help shape Dagger 2).
As people complained about that when trying to migrate from Dagger 1 to Dagger 2, I built Bullet as a proof-of-concept that it could be done in a third-party library; but I strongly believe that this is not the way to go (which also explains why I think it should not be built into Dagger 2).

So, back to code: you keep complaining about bad code and high coupling but haven't really showed what you mean by that, nor what would you think "good, loosely coupled" code would/should look like.

So let me throw a bunch of samples here as basis of discussion.

 1. Dagger 1 (or Bullet) with "micro framework" (abstract base class): loosely coupled and convenient, but "quality" is debatable:

    ```java
    abstract class Base {
        private void doInjection() {
          // some code to obtain an ObjectGraph
          graph.inject(this);
        }
    }
    ```

    ```java
    class Concrete extends Base {
      @Inject Dependency dep;
      // nothing special to do, everything's handled in the base class
    }
    ```

 2. Dagger 2, one component per class, instantiated within the class: high coupling. The exact opposite of the above.

    ```java
    class Concrete {
      @Component(…) interface MyComponent {
          void inject(Concrete self);
      }
      @Inject Dependency dependency;
      private void doInjection() {
         Dagger_Concrete_MyComponent.create().inject(this);
      }
    }
    ```

 3. Dagger 2, shared component, micro-framework: a middle-ground between 1 and 2. Best of both worlds, with a bit of boilerplate.

    ```java
    @Component
    interface MyComponent {
      void inject(Concrete concrete);
    }
    ```

    ```java
    abstract class Base {
      private void doInjection() {
        // some code to obtain a MyComponent instance
        doInjection(component);
      }
      protected abstract doInjection(MyComponent instance);
    }
    ```

    ```java
    class Concrete extends Base {
      @Inject Dependency dependency;
      @Override void doInjection(MyComponent component) {
        component.inject(this);
      }
    }
    ```

 4. Dagger 2, one component per class, instantiated within the class, but overridable by subclasses (e.g. during testing)

    ```java
    class Concrete {
      @Component(…) interface MyComponent {
          void inject(Concrete self);
      }
      @Inject Dependency dependency;
      protected MyComponent createMyComponent() {
        return Dagger_Concrete_MyComponent.create();
      }
      private void doInjection() {
        createMyComponent().inject(this);
      }
    }
    ```

 5. Dagger 2, one "injector" per class, grouped in a bigger component instantiated once for the app "somewhere": the class has no dependency on Dagger, code is self-contained.

    ```java
    class Concrete {
      interface Injector {
          void inject(Concrete self);
      }
      @Inject Dependency dependency;
      private void doInjection() {
        // some code to obtain an Injector instance (actually MyComponent instance)
        injector.inject(this);
      }
    }
    ```

    ```java
    @Component(…)
    interface MyComponent extends Concrete.Injector { }
    ```

 6. Dagger 2, "micro framework", one "injector" per class, grouped in a bigger component instantiated in a Base class; the class has no direct dependency on Dagger, but the base class has. You could imagine a variant where the base class gets the component instance from the "environment". Added complexity probably isn't worth it.

    ```java
    abstract class Base<T> {
      private final Class<T> injectorClass;
      protected Base(T injectorClass) { this.injectorClass = injectorClass; }
      private doInjection() {
        doInjection(createInjector());
      }
      protected T createInjector() {
        return injectorClass.cast(Dagger_MyComponent.create());
      }
      protected abstract void doInjection(T injector);
    }
    ```

    ```java
    class Concrete extends Base<Injector> {
      interface Injector {
          void inject(Concrete self);
      }
      public Concrete() { super(Injector.class); }
      @Inject Dependency dependency;
      @Override protected void doInjection(Injector injector) {
        injector.inject(this);
      }
    }
    ```

    ```java
    @Component(…)
    interface MyComponent extends Concrete.Injector { }
    ```

During testing, in the 4th and 6th samples, one could create a subclass that overrides the `createMyComponent()` or `createInjector()`method. This is enough for "unit" kind of tests, but not for a "testing flavor" of your application. You can then, however, replace your component or module or whatever with another one in your flavor: this is what build variants are for.
In all other samples (except the 2nd, shown here just for completeness), the graph and component could come from the "environment", so be configurable differently within tests or a "testing flavor" without touching the class under test directly or indirectly.
In the 5th and 6th samples, the `Injector` interface could be generated automatically (by an annotation processor) for any class that has an `@Inject` annotation on a field or method.

Now if you look at all those samples, the amount of boilerplate varies only by a couple of lines which most of the time could be generated by your IDE's snippets/templates feature.

**Comment by dalewking** [comment]
@tbroyer I will look at and respond to your examples later as I only have a few minutes right now. I can tell you that in my opinion if the class being injected or any of its superclasses have any reference to Dagger then it is incorrect because you have coupled that class to Dagger. Hiding it in a base class may make you feel better but does not remove the coupling.

I can at least contrast your examples with the sample using my library which is a rewrite of Bullet. For things like Activities there is one line added (see https://gitlab.com/NobleworksSoftware/NobleInjection/blob/master/githubclient-example/src/main/java/frogermcs/io/githubclient/ui/activities/RepositoriesListActivity.java):

        InjectionService.inject(this);

One static method call, that is all the boilerplate added (can't really get any less than that). And that static method call is completely agnostic of Dagger or whatever DI you want to use.

In the App startup you have to register an implementation of a simple interface:

    public interface Injector
    {
        <T> void inject(Context context, T target);
    }

The pattern I have used is to register an implementation that uses Dagger 2 wrapped with my Bullet-like wrapper for production but in unit tests I just have the test class implement the interface to insert the mocks. Here is the implementation of that interface for my sample app (which is a bit overcomplicated by going overboard to demonstrate using subcomponents): https://gitlab.com/NobleworksSoftware/NobleInjection/blob/master/githubclient-example/src/main/java/frogermcs/io/githubclient/di/GitHubClientInjector.java

The objects being injected are not coupled at all to Dagger. The only class in the system besides the modules and component interfaces that even reference Dagger is that one class that is only referenced by the App class.

Yes, it will fail at run time if I forgot to add an inject method to the component for an activity. But since I don't typically add activities to an app and ship them without running them at least once that generally is not much of a concern.

**Comment by dalewking** [comment]
Sample 1 is of course basically exactly what my proposal is. I do not like the base class having any knowledge of the Dagger graph and you did not show how the base class gets the graph, but that is easily abstracted away just as I showed so the base class can just have something like the InjectionService.inject(this) call I described in my implementation. I forgot to mention that this call can go in a base class so the boiler plate per injectable type can be zero

Sample 2 and 4 are awful and what I am strongly arguing against. Unit testing with 2 is impossible and you aren't even in my opinion doing dependency injection. Sadly, there are many examples given for Dagger 2 on Android that do this exactly. I hope everyone can agree that no one should be doing it like this. Requiring subclassing for testing is not an acceptable solution. What if the class I want to inject and test is final?

Sample 3 may be slightly less coupled, but without knowing how you propose getting the component it is hard to say for sure. There are plenty examples like this where basically they just moved the creation of the component from the Concrete class like in sample 2 to the Base class. That adds another link in the chain but is still as strongly coupled as sample 2 and suffers the same problems and in my opinion is still not doing DI since you cannot change what is being injected. You have to have a way that the base class can get a reference to an implementation of MyComponent without being in any way coupled to the actually generated component itself or preferably without any reference to Dagger at all. There are ways to do that, but I would also claim that since MyComponent will have a method for each type that need injected you have effectively coupled all injectable types to each other.

Sample 5 gets rid of the coupling together from every concrete type being tied directly to MyComponent. However, how does each component get the reference to the actual component and actually maintain this ideal of full graph analysis? I think if you actually try to implement that you will find that isn't as easy as you make it out to be as we will see with Sample 6. And this approach by the way is a crap ton of boiler plate to add to every single injectable type!

Sample 6 is actually no less coupled than sample 2 in my opinion since you are creating the component in the base class, but you could abstract that away from the base class so that you ask someone else for the MyComponent implementation, but guess what? Sample 6 does not have the full graph analysis that I am told is so important and the reason given for rejecting my proposal! If MyComponent does not actually implement Concrete.Injector it will fail at run-time not compile time. Sample 6 is basically an extremely verbose and complicated way to achieve what I am proposing with the same drawback of failure at run time.

So I find none of these Dagger 2 suggestions ]to be acceptable. They all introduce coupling and loads of tedious boilerplate code and/or make it impossible to do unit testing. Balancing those negatives against run-time failure of injection but with extremely clean, decoupled code, I choose the latter. 

But thank you very much for showing this range of possible solutions. I think it is very helpful to talk about it in real code. I have tried going down most of these roads (other than sample 2 which I rejected immediately) and found there is no acceptable answer in any of them. I had implemented something akin to sample 6 but it didn't actually achieve the goal of compile time validation of the graph, so I asked myself why I am adding all this boilerplate crap to all my classes and begun to question whether the problem was with the design of Dagger 2 itself and concluded that yes it was.

**Comment by ENikS** [comment]
+1 to @dalewking argument. 

@JakeWharton The reason no one is rushing to support this proposal is because not that many people are using Dagger 2 on Android yet. I think a lot of people will be looking for new DI solution once android bindings are released and adopted. When you do lookup by ID for your visuals adding few lines of boilerplate is trivial. But when your entire Activity code comes down to this:
```java

@Module
@ActivityScope
public class BootstrapperActivity extends AppCompatActivity {
...
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            ActivityComponent graph = DaggerActivityComponent.builder()
                    .bootstrapperActivity(this)
                    .build();

            MainActivityBinding binding = DataBindingUtil.setContentView(this,
                                                                         getProperLayout(...xyz...),
                                                                         graph);
           } catch (Exception e) {
            e.printStackTrace();
        }
    }
...
``` 
adding boilerplate is counterproductive. With this code I could have ALL of my activities to be implemented with just this bootstrapper onCreate and do all of the business logic in ViewModels.

**Comment by dalewking** [comment]
I realized I should probably expand on my complaints on the coupling of sample 3. For such a simple case where your app only has one component I'm sure many people would not care that much that all injectable types are tied to each other through that interface, but things quickly get more complicated when you start using multiple components or subcomponents. You then have multiple interfaces and how do you then do your base class? Do you end up with multiple base classes? So when you declare your injectable type you then have to choose the correct interface and the correct base class. You basically have to couple your class to the correct set of other classes.

And seriously imagine if you were using Dagger 1 and had an implementation like Sample 1 and decided to switch to Dagger 2. If you were told to implement any of samples 2-6 would you be saying, "wow, Dagger 2 is so much better than Dagger 1 and is making my life so much easier"? I can tell you when I switched our app to Dagger 2 with something akin to Sample 6 that was not the response from my team. The response from my team was a big WTF?!

**Comment by tbroyer** [comment]
> One static method call, that is all the boilerplate added

I'm sorry but relying on static state is nowhere close to being "good code" in my book, so I think I'll just unsubscribe from that thread before it turns into a really sterile discussion.
(I already disagree about what "constitutes" dependency injection)

> Sample 6 does not have the full graph analysis that I am told is so important

That reminds me I forgot to clarify what "full graph analysis" means: it means that starting from components, it can be verified statically at compile-time that all dependencies can be fulfilled (so you won't have "configuration errors" at runtime), i.e. the full dependency graph can be built at compile-time. It's a top-down validation (from components), not bottom-up (from injection sites).
In Dagger 1 you could have modules marked as libraries or incomplete and then forget to use a library to fulfill a missing dependency needed by an incomplete module.

Dagger 1 also doesn't limit the classes or objects you can ask from / pass to the `ObjectGraph` so you cannot tell at compile-time whether it's OK to use that or not; but this is a different issue from full graph analysis (I believe Jake misunderstood your proposal in his first two comments; third comment is about that other problem).

> If MyComponent does not actually implement Concrete.Injector it will fail at run-time not compile time.

Like all samples I've seen that cast the application into their own specific application class, or use a specific `getSystemService` (like u2020). Yes, there needs to be a cast somewhere, and yes it could fail at runtime (this is an issue with the Android SDK, not Dagger per se), but your proposal doesn't change that either:

> Yes, it will fail at run time if I forgot to add an inject method to the component for an activity. But since I don't typically add activities to an app and ship them without running them at least once that generally is not much of a concern.

One last thing that I don't think you explained is why what you're asking should be built into Dagger:

> The problem is that without access to the internals of the component it has to do some very hacky things that would not be necessary if this were built into the Dagger component and would be much more efficient.

I think that if you're really concerned about "efficiency" (which I read here mostly as runtime performance), then you can "ask" your components to only have methods that return `MethodInjector`s rather than also supporting methods taking an argument, and fail (or warn) otherwise.

**Comment by dalewking** [comment]
> I'm sorry but relying on static state is nowhere close to being "good code" in my book, so I think I'll just unsubscribe from that thread before it turns into a really sterile discussion.

@tbroyer That's pretty much the most ridiculous response I could imagine. You have a class instantiated by someone else with no references to anything else how the else do you think you can get a reference to something to do the injection without relying on some singleton state somewhere? It would be great if I could get that state injected, but since what I am trying to reference is the thing to do the injection it is a bit of a chicken and the egg problem. 

> That reminds me I forgot to clarify what "full graph analysis" means: it means that starting from components, it can be verified statically at compile-time that all dependencies can be fulfilled (so you won't have "configuration errors" at runtime), i.e. the full dependency graph can be built at compile-time. It's a top-down validation (from components), not bottom-up (from injection sites).
In Dagger 1 you could have modules marked as libraries or incomplete and then forget to use a library to fulfill a missing dependency needed by an incomplete module.

Well great then my proposal does NOT violate the full graph analysis. If you add an inject(Foo) method to the component Dagger would still do all the same graph analysis and still guarantee that its dependencies can be fulfilled. My proposal just adds another way to ask for injections. However I doubt that Jake's definition agrees with yours.

> Dagger 1 also doesn't limit the classes or objects you can ask from / pass to the ObjectGraph so you cannot tell at compile-time whether it's OK to use that or not; but this is a different issue from full graph analysis (I believe Jake misunderstood your proposal in his first two comments; third comment is about that other problem).

Jake often misunderstands things. He doesn't always read what others are saying as he admitted in an issue in another library where he was wrong and disagreeing with me, "I just never really read what you had (still haven't and full disclosure I didn't read your response)". The way Jake pretty much automatically treats everyone who posts into a GitHub issue like they are an idiot is actually a running joke among my dev team. With one lone exception that is how every interaction I have had with Jake has turned out.

> Like all samples I've seen that cast the application into their own specific application class, or use a specific getSystemService (like u2020). Yes, there needs to be a cast somewhere, and yes it could fail at runtime (this is an issue with the Android SDK, not Dagger per se), but your proposal doesn't change that either:

Not sure I would call it an issue with the Android SDK other than that it creates the important objects for you, but as you said the same applies to Servlets.

> One last thing that I don't think you explained is why what you're asking should be built into Dagger:
>
> I think that if you're really concerned about "efficiency" (which I read here mostly as runtime performance), then you can "ask" your components to only have methods that return MethodInjectors rather than also supporting methods taking an argument, and fail (or warn) otherwise.

And that is what my rewrite of your Bullet library does. If you have any classes that you only declare in the component as inject methods then it does warn you.

Yes, efficiency is part of it, but one of the main reasons it should be part of Dagger 2 is that basically Dagger 2, particularly on Android is dying. I can tell you anecdotally that my boss has said that any new projects will use Dagger 1, not Dagger 2, and he claims that most people are passing on Dagger 2. Those that are trying to use it are they are doing it in ways that destroy most of the benefits of DI (variations on sample 2 and 3).

Making it an official part of Dagger 2 would help Dagger 2 to be able to grow and to improve the code that is out there, better than some third party add-on library ever could. The ability to get people to even understand the issue enough to know about the 3rd party library and to use it is limited. Just look how much trouble I have had even getting people in this issue who are familiar with Dagger to understand it.

Efficiency in terms of speed is one thing and why I rejected your Bullet library. There is also efficiency in terms of number of objects created at runtime and number of lines of code. The other thing I worry about is how straightforward the implementation is as that was also a design goal of Dagger 2.

If you want an example for a simple component like this:

    @Component
    interface SimpleComponent {
        void inject(B b);
        void inject(A a);
        void inject(C c);
    }

This is what my library generates, which is the most efficient algorithm I could come up with which relies on the Generic component base class I linked to earlier:

    public final class GenericSimpleComponent extends GenericComponent<SimpleComponent>
    {
        public GenericSimpleComponent(SimpleComponent component)
        {
            super(component);
        }
    
        protected ClassInjectorMap buildInjectorMap()
        {
            ClassInjectorMap map = new ClassInjectorMap(5);
            map.put(A.class, new SimpleInjector(0));
            map.put(B.class, new SimpleInjector(1));
            map.put(C.class, new SimpleInjector(2));
            return map;
        }
    
        private class SimpleInjector<T> implements MembersInjector<T>
        {
            private final int index;
    
            private SimpleInjector(int index)
            {
                this.index = index;
            }
    
            public void injectMembers(T instance)
            {
                switch (index)
                {
                    case 0:
                        component.inject((A) instance);
                        break;
                    case 1:
                        component.inject((B) instance);
                        break;
                    case 2:
                        component.inject((C) instance);
                        break;
                    default:
                        throw new IllegalArgumentException("Invalid Index!");
                }
            }
        }
    }

If this were part of Dagger itself you wouldn't use the base class but would instead redirect to a utility class in the dagger runtime. All that would go in the component would be the generic injection method redirecting to the utility class and something like the buildInjectorMap method but since you are in the component itself and have access to all of the member injector member variables, it would just be like this:

        ClassInjectorMap buildInjectorMap()
        {
            ClassInjectorMap map = new ClassInjectorMap(5);
            map.put(A.class, aMembersInjector);
            map.put(B.class, bMembersInjector);
            map.put(C.class, cMembersInjector);
            return map;
        }

**Comment by cgruber** [comment]
Please desist from commentary on personalities and other personal judgments
on our issues.  Technical and produced discussions can be vigorous.  That
does not give people a pass to complain about individuals.  Take that to
social media.

**Comment by ENikS** [comment]
Implementation of base injector similar to **SimpleInjector** could be done lazily so it does not consume any resources or CPU if it is not required. Also, if dictionary/map is initialized at runtime, class names will be adequate because even obfuscated it would still represent unique and distinct identifier. No need for fancy index juggling, long **switch** of **instanceOf** statements.

We already determined that implementation of this functionality during compile step is trivial. The big question is whether Google is willing to listen to this feedback and augment its core design principal of Dagger 2. 

Does anyone from Google care to comment?

**Comment by dalewking** [comment]
> Also, if dictionary/map is initialized at runtime, class names will be adequate because even obfuscated it would still represent unique and distinct identifier. No need for fancy index juggling, long switch of instanceOf statements.

Yes, the hacky index juggling is only needed in my 3rd party library because I don't have access to the MembersInjector instances from the component unless the component is declared to give access (uses the methods like MembersInjector<Foo> getFooInjector() and not just methods like Foo inject(Foo)). If implemented inside of the component the switch and SimpleInjector are totally not necessary, thus part of my statements about how it would be more efficient inside of Dagger.

There is no reason to use class name for the map instead of the actual class. It would only survive obfuscation if it were something like:

    injectorMap.put(A.class.getCanonicalName(), aMembersInjector);

But why use the name when you can use the class itself?

**Comment by ENikS** [comment]
@dalewking These are all valid but irrelevant points at the moment. 

I think "fundamental" question whether "core principle" should be changed have to be decided first.

**Comment by JakeWharton** [comment]
> LOL! You aren't using Dagger 2!

News to me! Anything else I need to know about myself?

> You ARE using Dagger 1 which has this feature so you are currently using this feature.

Right. Because the desired feature of full-graph analysis and verification at compile time doesn't exist. Because that was a design principle of Dagger 2.

It's like you're not even reading what I'm writing. We have a joke in our dev team...

> It encourages tight coupling to the component itself

Only if _you_ let it.

Each activity can expose
```java
interface Injector {
  void inject(ThisActivity activity);
}
```
which is implemented on the injector (which is still type-safe unlike your proposal) and then use `getSystemService` as the lookup mechanism.

This is the same, well-established pattern of fragments calling up to their activities via a defined contract interface. Why should this be any different?

**Comment by dalewking** [comment]
> >  LOL! You aren't using Dagger 2!

> News to me! Anything else I need to know about myself?

Apparently you need to know that someone has hacked your account and in the issue requesting Dagger 2 to be added to u2020 made it seem you said things like: "Right now we're using Dagger 1 internally and probably not upgrading any time soon", "It [Dagger 1] serves our use case much better than Dagger 2"

> Right. Because the desired feature of full-graph analysis and verification at compile time doesn't exist. Because that was a design principle of Dagger 2.

And the definition of full graph analysis given by tbroyer does not go away with my proposal. I'm not sure if your definition is the same as his. It appears your definition is that injection cannot fail at runtime.

> > It encourages tight coupling to the component itself

> Only if you let it.

And we are still looking for the answer for how you don't let it AND only ever fail at compile time.

> Each activity can expose

>    interface Injector {
>      void inject(ThisActivity activity);
>    }

Which is sample 5/6 from @tbroyer but you left out the additional boilerplate that has to accompany this:

  @Override protected void doInjection(Injector injector) {
    injector.inject(this);
  }

> which is implemented on the injector (which is still type-safe unlike your proposal) and then use getSystemService as the lookup mechanism.

Wrong! What is the getSystemService going to do to provide the instance of the interface? Just like what TBroyer did in Sample 6 it will end up just doing a cast which will not be checked at compile time and will only fail at run time. So it is no more type-safe than my proposal at the expense of adding a ton of boilerplate to each injectable type, which I would say encourages people to just go with tight coupling since it is a lot less boilerplate.

If you have some other magical implementation that maintains the type safety and the loose coupling, I'm all ears, but as far as I know none exists.

> This is the same, well-established pattern of fragments calling up to their activities via a defined contract interface. Why should this be any different?

It isn't any different because it is not type-safe either and will fail at runtime. So if you are only going to be able to fail at runtime why add a bunch of boilerplate code that makes you no safer?

**Comment by JakeWharton** [comment]
> you said things like: "Right now we're using Dagger 1 internally and probably not upgrading any time soon", "It [Dagger 1] serves our use case much better than Dagger 2"

Fun fact: I work on more than u2020 and a single internal app.

> it will end up just doing a cast which will not be checked at compile time and will only fail at run time.

> So if you are only going to be able to fail at runtime why add a bunch of boilerplate code that makes you no safer?

Because it doesn't live in Dagger and doesn't undermine its principles.

**Comment by gk5885** [comment]
This conversation seems to have gone a bit off the rails.  Let's all take a moment to take a break from this and collect our thoughts.

@netdpb and I will sit down and distill what we can from this thread and report back with how we plan to integrate the feedback into our roadmap.  At that time, I'll open the conversation back up again and we can get back onto a productive track.
